Mathlib/CategoryTheory/Functor/KanExtension/Pointwise.lean

```diff
@@ -0,0 +1,191 @@
+`E.coconeAt Y` is a colimit cocone. -/
+def IsPointwiseLeftKanExtensionAt (Y : D) := IsColimit (E.coconeAt Y)
+
+variable {E}
+
+lemma IsPointwiseLeftKanExtensionAt.hasPointwiseLeftKanExtensionAt
+    {E : LeftExtension L F} {Y : D} (h : E.IsPointwiseLeftKanExtensionAt Y) :
+  HasPointwiseLeftKanExtensionAt L F Y := ⟨_, h⟩
+
+variable (E)
+
+/-- A left extension `E : LeftExtension L F` is a pointwise left Kan extension when
+it is a pointwise left Kan extension at any object. -/
+abbrev IsPointwiseLeftKanExtension := ∀ (Y : D), E.IsPointwiseLeftKanExtensionAt Y
+
+variable (h : E.IsPointwiseLeftKanExtension)
+
+lemma IsPointwiseLeftKanExtension.hasPointwiseLeftKanExtension :
+    HasPointwiseLeftKanExtension L F :=
+  fun Y => (h Y).hasPointwiseLeftKanExtensionAt
+
+/-- A pointwise left Kan extension is universal, i.e. it is a left Kan extension. -/
+def IsPointwiseLeftKanExtension.isUniversal : E.IsUniversal :=
+  IsInitial.ofUniqueHom (fun G => StructuredArrow.homMk
+        { app := fun Y => (h Y).desc (LeftExtension.coconeAt G Y)
+          naturality := fun Y₁ Y₂ φ => (h Y₁).hom_ext (fun X => by
+            rw [(h Y₁).fac_assoc (coconeAt G Y₁) X]
+            simpa using (h Y₂).fac (coconeAt G Y₂) ((CostructuredArrow.map φ).obj X)) }
+      (by
+        ext X
+        simpa using (h (L.obj X)).fac (LeftExtension.coconeAt G _) (CostructuredArrow.mk (𝟙 _))))
+    (fun G => by
+      suffices ∀ (m₁ m₂ : E ⟶ G), m₁ = m₂ by intros; apply this
+      intro m₁ m₂
+      ext Y
+      apply (h Y).hom_ext
+      intro X
+      have eq₁ := congr_app (StructuredArrow.w m₁) X.left
+      have eq₂ := congr_app (StructuredArrow.w m₂) X.left
+      dsimp at eq₁ eq₂ ⊢
+      simp only [assoc, NatTrans.naturality]
+      rw [reassoc_of% eq₁, reassoc_of% eq₂])
+
+lemma IsPointwiseLeftKanExtension.isLeftKanExtension :
+    E.right.IsLeftKanExtension E.hom where
+  nonempty_isUniversal := ⟨h.isUniversal⟩
+
+lemma IsPointwiseLeftKanExtension.hasLeftKanExtension :
+    HasLeftKanExtension L F :=
+  have := h.isLeftKanExtension
+  HasLeftKanExtension.mk E.right E.hom
+
+end LeftExtension
+
+section
+
+variable [HasPointwiseLeftKanExtension L F]
+
+/-- The constructed pointwise left Kan extension when `HasPointwiseLeftKanExtension L F` holds. -/
+@[simps]
+noncomputable def pointwiseLeftKanExtension : D ⥤ H where
+  obj Y := colimit (CostructuredArrow.proj L Y ⋙ F)
+  map {Y₁ Y₂} f :=
+    colimit.desc (CostructuredArrow.proj L Y₁ ⋙ F)
+      (Cocone.mk (colimit (CostructuredArrow.proj L Y₂ ⋙ F))
+        { app := fun g => colimit.ι (CostructuredArrow.proj L Y₂ ⋙ F)
+            ((CostructuredArrow.map f).obj g)
+          naturality := fun g₁ g₂ φ => by
+            simpa using colimit.w (CostructuredArrow.proj L Y₂ ⋙ F)
+              ((CostructuredArrow.map f).map φ) })
+  map_id Y := colimit.hom_ext (fun j => by
+    dsimp
+    simp only [colimit.ι_desc, comp_id]
+    congr
+    apply CostructuredArrow.map_id)
```

Comment:
I don't love that all of these proofs need to argue using equality of objects, but I don't see an easy way to avoid this.

Reply:
Yes, I was feeling the same, but I do not see a nice way to create the suitable goals for using the naturality with respect to an iso rather than `congr` followed by an equality.